src/khoj/database/models/__init__.py

```diff
@@ -84,23 +85,33 @@ class ModelType(models.TextChoices):
 
 
 class Agent(BaseModel):
-    creator = models.ForeignKey(
-        KhojUser, on_delete=models.CASCADE, default=None, null=True, blank=True
-    )  # Creator will only be null when the agents are managed by admin
+    creator = models.ForeignKey(KhojUser, on_delete=models.CASCADE, default=None, null=True, blank=True)
     name = models.CharField(max_length=200)
     personality = models.TextField()
     avatar = models.URLField(max_length=400, default=None, null=True, blank=True)
     tools = models.JSONField(default=list)  # List of tools the agent has access to, like online search or notes search
     public = models.BooleanField(default=False)
     managed_by_admin = models.BooleanField(default=False)
     chat_model = models.ForeignKey(ChatModelOptions, on_delete=models.CASCADE)
+    slug = models.CharField(max_length=200, default=None, null=True, blank=True)
 
 
 @receiver(pre_save, sender=Agent)
-def check_public_name(sender, instance, **kwargs):
-    if instance.public:
+def verify_agent(sender, instance, **kwargs):
+    # check if this is a new instance
+    if instance._state.adding:
         if Agent.objects.filter(name=instance.name, public=True).exists():
             raise ValidationError(f"A public Agent with the name {instance.name} already exists.")
+        if Agent.objects.filter(name=instance.name, creator=instance.creator).exists():
+            raise ValidationError(f"A private Agent with the name {instance.name} already exists.")
+
+        slug = instance.name.lower().replace(" ", "-")
+        observed_random_numbers = set()
+        while Agent.objects.filter(slug=slug).exists():
+            random_number = choice([i for i in range(0, 10000) if i not in observed_random_numbers])
+            observed_random_numbers.add(random_number)
+            slug = f"{slug}-{random_number}"
+        instance.slug = slug
```

Comment:
Can we simplify this to use `random.randint(10000)`?

```suggestion
        slug = instance.name.lower().replace(" ", "-")
        random_numbers = random.randint(10000)
        slug = f"{slug}-{random_number}"
        instance.slug = slug
```

Reply:
This wouldn't work in the case that there's a collision. That's why I'm using the `choice` function while extracting the values that are already seen.